Cocotb abnormal exit is ignored

Hi,

The below code expects the test to fail.
However, veryl command actually reports a success.


code:
```systemverilog
module A (
    a: input logic<1>,
) {
    let b: logic<2> = a; // this line should cause a mismatch-width error
}

#[test(test, A)]
embed (cocotb) py{{{
import cocotb

@cocotb.test()
async def test(dut):
    assert False, "this test is expected to fail"
}}}
```

command: `veryl test`


output:
```
[INFO ]    Executing test (test)
[INFO ]    Succeeded test (test)
[INFO ]    Completed tests : 1 passed, 0 failed
```

As far as I can tell, this is due to an error in the build phase by Verilator, and `runner.py` exits with exit code 1, which is ignored.

I added exit code check and stderr parser for cocotb.

```
[INFO ]    Executing test (test)
/tmp/.tmpShr0Br/runner.py:3: UserWarning: Python runners and associated APIs are an experimental feature and subject to change.
  import cocotb.runner
%Warning-WIDTHEXPAND: /home/hatta/work/repos/veryl/temp/test.sv:5:36: Operator ASSIGN expects 2 bits on the Assign RHS, but Assign RHS's VARREF 'a' generates 1 bits.
                                                                    : ... note: In instance 'temp_A'
    5 |     logic [2-1:0] b; always_comb b = a;
      |                                    ^
                      ... For warning description see https://verilator.org/warn/WIDTHEXPAND?v=5.024
                      ... Use "/* verilator lint_off WIDTHEXPAND */" and lint_on around source to disable this message.
%Error: Exiting due to 1 warning(s)
Process 'perl' terminated with error 1
[ERROR]       cocotb failed by error code exit status: 1
[ERROR]       Failed test (test)
[ERROR]    Completed tests : 0 passed, 1 failed
```

--- a/crates/veryl/src/runner/cocotb.rs
+++ b/crates/veryl/src/runner/cocotb.rs
@@ -44,43 +44,47 @@ impl Cocotb {
         Box::new(self) as Box<dyn Runner>
     }
 
-    fn parse_line(&mut self, line: &str) {
+    fn parse_line(&mut self, line: &str, force_error: bool) {
         self.debug(line);
 
-        if !line.starts_with("                ") {
-            self.state = State::Idle;
-        }
+        if force_error {
+            self.error(line);
+        } else {
+            if !line.starts_with("                ") {
+                self.state = State::Idle;
+            }
 
-        match self.state {
-            State::Idle => {
-                if line.ends_with("failed") {
-                    self.error(line);
-                    self.state = State::Error;
-                } else if line.starts_with("     0.00ns INFO") {
-                    self.info(line);
-                    self.state = State::Info;
-                } else if line.starts_with("     0.00ns WARNING") {
-                    self.warning(line);
-                    self.state = State::Warning;
-                } else if line.starts_with("     0.00ns ERROR") {
-                    self.error(line);
-                    self.state = State::Error;
-                } else if line.starts_with("     0.00ns CRITICAL") {
-                    self.fatal(line);
-                    self.state = State::Fatal;
+            match self.state {
+                State::Idle => {
+                    if line.ends_with("failed") {
+                        self.error(line);
+                        self.state = State::Error;
+                    } else if line.starts_with("     0.00ns INFO") {
+                        self.info(line);
+                        self.state = State::Info;
+                    } else if line.starts_with("     0.00ns WARNING") {
+                        self.warning(line);
+                        self.state = State::Warning;
+                    } else if line.starts_with("     0.00ns ERROR") {
+                        self.error(line);
+                        self.state = State::Error;
+                    } else if line.starts_with("     0.00ns CRITICAL") {
+                        self.fatal(line);
+                        self.state = State::Fatal;
+                    }
                 }
-            }
-            State::Info => {
-                self.info(line);
-            }
-            State::Warning => {
-                self.warning(line);
-            }
-            State::Error => {
-                self.error(line);
-            }
-            State::Fatal => {
-                self.fatal(line);
+                State::Info => {
+                    self.info(line);
+                }
+                State::Warning => {
+                    self.warning(line);
+                }
+                State::Error => {
+                    self.error(line);
+                }
+                State::Fatal => {
+                    self.fatal(line);
+                }
             }
         }
     }
@@ -90,7 +94,19 @@ impl Cocotb {
         let mut reader = FramedRead::new(stdout, LinesCodec::new());
         while let Some(line) = reader.next().await {
             let line = line.into_diagnostic()?;
-            self.parse_line(&line);
+            self.parse_line(&line, false);
+        }
+
+        let ecode = child.wait().await.into_diagnostic()?;
+        if !ecode.success() {
+            let stderr = child.stderr.take().unwrap();
+            let mut reader = FramedRead::new(stderr, LinesCodec::new());
+            while let Some(line) = reader.next().await {
+                let line = line.into_diagnostic()?;
+                self.parse_line(&line, true);
+            }
+            error!("cocotb failed by error code {ecode}");
+            self.failure();
         }
         Ok(())
     }
@@ -185,7 +201,7 @@ runner.test(
                 .arg("runner.py")
                 .current_dir(temp_dir.path())
                 .stdout(Stdio::piped())
-                .stderr(Stdio::null())
+                .stderr(Stdio::piped())
                 .spawn()
                 .into_diagnostic()
                 .wrap_err("Failed to run \"python3\"")?;

